espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/XMLHelpers.kt

```diff
@@ -21,36 +21,43 @@ import java.util.regex.Pattern
 object XMLHelpers {
     // XML 1.0 Legal Characters (http://stackoverflow.com/a/4237934/347155)
     // #x9 | #xA | #xD | [#x20-#xD7FF] | [#xE000-#xFFFD] | [#x10000-#x10FFFF]
-    private val XML10_PATTERN = Pattern.compile("[^" + "\u0009\r\n" +
-            "\u0020-\uD7FF" + "\uE000-\uFFFD" + "\ud800\udc00-\udbff\udfff" + "]")
+    private val XML10_PATTERN = Pattern.compile(
+        "[^" + "\u0009\r\n" +
+                "\u0020-\uD7FF" + "\uE000-\uFFFD" + "\ud800\udc00-\udbff\udfff" + "]"
+    )
+
     // https://stackoverflow.com/questions/3158274/what-would-be-a-regex-for-valid-xml-names
-    private val XML10_START_TAG_PATTERN = Pattern.compile("^[^" + "_" +
-            "A-Z" + "a-z" + "\u00C0-\u00D6" + "\u00F8-\u02FF" +
-            "\u200C-\u200D" + "\u2070-\u218F" + "\u2C00-\u2FEF" +
-            "\u3001-\uD7FF" + "\uF900-\uFDCF" + "\uFDF0-\uFFFD" +
-            "\ud800\udc00-\udbff\udfff" + "]+")
-    private val XML10_TAG_PATTERN = Pattern.compile("[^" + "\\-._" +
-            "0-9" + "\u00B7" + "\u0300-\u036F" + "\u203F-\u2040" +
-            "A-Z" + "a-z" + "\u00C0-\u00D6" + "\u00F8-\u02FF" +
-            "\u200C-\u200D" + "\u2070-\u218F" + "\u2C00-\u2FEF" +
-            "\u3001-\uD7FF" + "\uF900-\uFDCF" + "\uFDF0-\uFFFD" +
-            "\ud800\udc00-\udbff\udfff" + "]")
+    private val XML10_START_TAG_PATTERN = Pattern.compile(
+        "^[^" + "_" +
+                "A-Z" + "a-z" + "\u00C0-\u00D6" + "\u00F8-\u02FF" +
+                "\u200C-\u200D" + "\u2070-\u218F" + "\u2C00-\u2FEF" +
+                "\u3001-\uD7FF" + "\uF900-\uFDCF" + "\uFDF0-\uFFFD" +
+                "\ud800\udc00-\udbff\udfff" + "]+"
+    )
+    private val XML10_TAG_PATTERN = Pattern.compile(
+        "[^" + "\\-._" +
+                "0-9" + "\u00B7" + "\u0300-\u036F" + "\u203F-\u2040" +
+                "A-Z" + "a-z" + "\u00C0-\u00D6" + "\u00F8-\u02FF" +
+                "\u200C-\u200D" + "\u2070-\u218F" + "\u2C00-\u2FEF" +
+                "\u3001-\uD7FF" + "\uF900-\uFDCF" + "\uFDF0-\uFFFD" +
+                "\ud800\udc00-\udbff\udfff" + "]"
+    )
 
     fun toNodeName(str: String): String {
         val nodeName = XML10_START_TAG_PATTERN
-                .matcher(str)
-                .replaceAll("")
+            .matcher(str)
+            .replaceAll("")
         return XML10_TAG_PATTERN
-                .matcher(nodeName)
-                .replaceAll("_")
+            .matcher(nodeName)
+            .replaceAll("_")
     }
 
     fun toSafeString(source: Any?, replacement: String): String? {
         return if (source == null)
-            null
+            ""
```

Comment:
this change might break existing espresso stuff. Please verify XML files generated by Espresso did not change

Reply:
It does change, Instead of no nodes, it adds nodes with empty values like `text="" `  in the XML.

Reply:
@KazuCocoa How do you think, would it be ok if we change that behaviour (no attribute vs empty-string attribute for null values) now or it would be better to keep it as it was for espresso?

Reply:
mm, empty-string can consider the element is set as empty-string, so it possibly confuses users if the attribute is set as empty or does not exist. btw, then, can XPath locator find the element with an empty string `text`?

I'd prefer to keep the previous case (null) to explicit no-attributes

Reply:
btw in uiautomator2 and xcuitest we do transform nulls to empty strings

Reply:
oh, i see...
So, then, are https://github.com/appium/appium-uiautomator2-driver/#element-attributes empty strings when they have no values? (I think we should address this point.)

Then an empty string is fine, but I think it would be nice to create another PR to leave the breaking change as a separate message. (appium 1.22 has been released, so we will not have such a central changelog for now.)

Reply:
basically, how it works there, in the xml we replace null values with empty strings, but for `getAttribute` call nulls are still allowed

Reply:
okay, it makes sense to me > but for getAttribute call nulls are still allowed

I see the replacement.
I'm ok to change this behavior to make the same as other drivers. But I still would like to make this change as a separate one from this compose feature.

Reply:
I think the same. In fact, we should reconsider changing the behavior of other drivers (no text is not the same as empty text)
For now, it is best to keep the compose behavior in line with espresso to avoid detour